crates/sdk/src/masp.rs

```diff
@@ -491,15 +540,62 @@ pub fn find_valid_diversifier<R: RngCore + CryptoRng>(
 }
 
 /// Determine if using the current note would actually bring us closer to our
-/// target
-pub fn is_amount_required(src: I128Sum, dest: I128Sum, delta: I128Sum) -> bool {
-    let gap = dest - src;
+/// target. Returns the unused amounts (change) of delta if any
+pub fn is_amount_required(
+    src: I128Sum,
+    dest: I128Sum,
+    normed_delta: I128Sum,
+    opt_delta: Option<I128Sum>,
+) -> Option<I128Sum> {
+    let mut changes = None;
+    let gap = dest.clone() - src;
+
     for (asset_type, value) in gap.components() {
-        if *value > 0 && delta[asset_type] > 0 {
-            return true;
-        }
+        if *value > 0 && normed_delta[asset_type] > 0 {
+            let signed_change_amt =
+                checked!(normed_delta[asset_type] - *value).unwrap_or_default();
+            let unsigned_change_amt = if signed_change_amt > 0 {
+                signed_change_amt
+            } else {
+                // Even if there's no change we still need to set the return
+                // value of this function to be Some so that the caller sees
+                // that this note should be used
+                0
+            };
+
+            let change_amt = I128Sum::from_nonnegative(
+                asset_type.to_owned(),
+                unsigned_change_amt,
+            )
+            .expect("Change is guaranteed to be non-negative");
+            changes = changes
+                .map(|prev| prev + change_amt.clone())
```

Comment:
curious that unchecked arithmetic is ok here

Reply:
we don't have the strict clippy checks in the sdk :/ - ideally, we shouldn't be implementing any logic here

Reply:
imho we should move the sdk masp mod into the shielded_token crate

Reply:
It's actually checked if I recall correctly